File: haskell/sender.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Control.Concurrent
import Control.Monad.Cont

import qualified Data.ByteString.UTF8 as U8B
import qualified Data.ByteString as B
import Network.Socket
import qualified Network.Socket.ByteString as B

addr = tupleToHostAddress (127, 0, 0, 1)
port = 4321

main = withSocketsDo $ do
    -- Create the socket
    sock <- socket AF_INET Datagram 0

    -- Send a datagram
    let sa = SockAddrInet port addr
    putStrLn "Sending 3 messages:"
    forM_ [1,2,3] $ \i -> do
        let msg = "Hi, there " ++ show i
        putStrLn $ "  Sending message \"" ++ msg ++ "\""
        B.sendTo sock (U8B.fromString msg) sa
        threadDelay 1000000
    putStrLn "Done."

    -- Close the socket
    close sock

File: python/sender.py
#!/usr/bin/env python3

import socket
import time

addr = '127.0.0.1'
port = 4321

socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, 0)

print("Sending 3 messages:")
for i in range(1, 4):
    msg = f"Hi, there {i}"
    print(f"  Sending message \"{msg}\"")
    socket.sendto(msg.encode('utf-8'), (addr, port))
    time.sleep(1)
print("Done.")
socket.close()
